src/core/server/saved_objects/migrations/types.ts

```diff
@@ -39,10 +39,10 @@ import { SavedObjectsMigrationLogger } from './core/migration_logger';
  *
  * @public
  */
-export type SavedObjectMigrationFn = (
-  doc: SavedObjectUnsanitizedDoc,
-  context: SavedObjectMigrationContext
-) => SavedObjectUnsanitizedDoc;
+export type SavedObjectMigrationFn<
+  TInputDoc extends SavedObjectUnsanitizedDoc | SavedObjectSanitizedDoc = SavedObjectUnsanitizedDoc,
```

Comment:
I think we need to add two generics cause it allow us to avoid a lot of casting in future. 

Like e.g. here:  `src/plugins/dashboard/server/saved_objects/dashboard_migrations.ts`
Also using this approach we can guarantee that all Doc's extended from `SavedObjectUnsanitizedDoc` or `SavedObjectSanitizedDoc`

Reply:
To follow our slack discussion:

This is an alternative of https://github.com/elastic/kibana/pull/63943

However on #63943 we choose to only generify the `attributes` property of `SavedObjectUnsanitizedDoc`, as we didn't see any (valid) reasons plugin migration functions should be able to superseed/override the root SO type. This could also be dangerous as it may cause developer errors trying to push changes/properties not defined in the schema.

Do you have any usages where you would need to use type override on the whole `SavedObjectUnsanitizedDoc` type?

Reply:
At least in version 7.0.0 `references` were added. And we should somehow handle it.
Also who knows what will be added in future?

From the other hand if someone wants to add something into doc we have at least 2 options: use JS without typings, just cast object to needed type =) So, validation should be implemented in a different way

Reply:
Currently there's an index signature so anything goes, but we should definitely remove that https://github.com/elastic/kibana/blob/97d1685c3dea682f80fd1a907bbc1d6f3702ea85/src/core/server/saved_objects/serialization/types.ts#L63

Root properties cannot be added by plugins, so if we ever add them in Core, we can update this type at the same time like we recently did for the new `namespaces` root property.

Reply:
@alexwizp could I get your review on https://github.com/elastic/kibana/pull/63943? Any argument on why the PR wouldn't answer the need in a better/safer way is welcome.